Add MINP for soft motors

tboegi opened the pull request:
Make it possible to set bits of the MSTA field for soft motors. Add a new input link, called MINP, beside the existing DINP and RINP

MINP will read the new MSTA bits, but filter out the RA_DONE bit. The RA_DONE bit will be taken from DINP (as before) and the RA_DONE bit inside MINP will be ignored, preserving the RA_DONE coming from DINP

Note that the MINP needs to follow the bit-definition in MSTA. It may update RA_PLUS_LS, RA_MINUS_LS.
RA_HOMED or other bits may be set as well, whatever the application needs. MINP may be pointed to the output of a calc or transform record.

kmpeters commented:
@tboegi, I switched the master branch to use the sequencer mirror.  If you rebase this pull request on the master branch, the CI builds should succeed.

tboegi committed:
Add MINP for soft motors

Make it possible to set bits of the MSTA field for soft motors.
Add a new input link, called MINP, beside the existing DINP and RINP

MINP will read the new MSTA bits, but filter out the RA_DONE bit.
The RA_DONE bit will be taken from DINP (as before) and the RA_DONE bit
inside MINP will be ignored, preserving the RA_DONE coming from DINP

Note that the MINP needs to follow the bit-definition in MSTA.
It may update RA_PLUS_LS, RA_MINUS_LS.
RA_HOMED or other bits may be set as well, whatever the application needs.
MINP may be pointed to the output of a calc or transform record.

tboegi commented:
2023-12-11T07:45:00.0015590Z   File "/Users/runner/work/motor/motor/.ci/cue.py", line 14, in <module>
2023-12-11T07:45:00.0016620Z     import distutils.util
2023-12-11T07:45:00.0018040Z ModuleNotFoundError: No module named 'distutils'
2023-12-11T07:45:00.0148950Z ##[error]Process completed with exit code 1.

tboegi commented:
@kmpeters I am confused as well. More digging needed. 
The code was compiled (and tested) locally running on a Intel-Mac.

kmpeters commented:
Pedro Nariyoshi reports that these changes work well in his test setup:

https://epics.anl.gov/tech-talk/2023/msg01901.php

kmpeters commented:
I'm testing this pull request with using the following databases in the motorMotorSim example IOC, which implements the soft motor example from the motor documentation and links an mbboDirect record to the MINP field of the soft motor:

[SoftMotorExample.db.txt](https://github.com/epics-modules/motor/files/13687222/SoftMotorExample.db.txt)
[SoftMSTA.db.txt](https://github.com/epics-modules/motor/files/13687223/SoftMSTA.db.txt)

I'm using the first simulated motor as the "real" motor:
```
dbLoadRecords("$(TOP)/motorSimApp/Db/SoftMotorExample.db", "P=$(PREFIX),M=m1,SM=sm1")
dbLoadRecords("$(TOP)/motorSimApp/Db/SoftMSTA.db", "P=$(PREFIX),SM=sm1")
```
I'm able to set the high hardware limit by doing this: `caput IOC:sm1:msta.B2 1`
I'm able to set the low hardware limit by doing this: `caput IOC:sm1:msta.BD 1`

I observe the following:

1. Setting hardware limits via the MINP field causes the soft motor to think the motor is done moving shortly after the move is initiated (the target position gets synced with the RBV field), however the real motor is never stopped.
2. The soft motor's MSTA field isn't set correctly when the homed bit is set, `caput IOC:sm1:msta.BF 1`.  The `IOC:sm1:msta.VAL` is `32768`, but the soft motor's MSTA field has a value of `0xffff8002`.

tboegi commented:
About the STOP: I need to check, if we want to have this functionality ?

@kmpeters The homed bit is bit 14, not 15.

`caput IOC:sm1:msta.BE 1
`

14        unsigned int RA_HOMED       :1; /* Axis has been homed.*/
13        unsigned int RA_MINUS_LS    :1; /* minus limit switch has been hit */
12        unsigned int CNTRL_COMM_ERR :1; /* Controller communication error. */
11        unsigned int GAIN_SUPPORT   :1; /* Motor supports closed-loop position control. */
10        unsigned int RA_MOVING      :1; /* non-zero velocity present */
9         unsigned int RA_PROBLEM     :1; /* driver stopped polling */
8         unsigned int EA_PRESENT     :1; /* encoder is present */
7         unsigned int EA_HOME        :1; /* encoder home signal on */
6         unsigned int EA_SLIP_STALL  :1; /* slip/stall detected */
5         unsigned int EA_POSITION    :1; /* position maintenence enabled */
4         unsigned int EA_SLIP        :1; /* encoder slip enabled */
3         unsigned int RA_HOME        :1; /* The home signal is on */
2         unsigned int RA_PLUS_LS     :1; /* plus limit switch has been hit */
1         unsigned int RA_DONE        :1; /* a motion is complete */
0         unsigned int RA_DIRECTION   :1; /* (last) 0=Negative, 1=Positive */

tboegi commented:
@kmpeters 
I just realize that the .MSTA field is declared as 
        field(MSTA,DBF_ULONG) {
While we see this:
void soft_minp_func(struct motorRecord *mr, short newminp)


So I think that the PR needs improvements ?!

kmpeters commented:
> About the STOP: I need to check, if we want to have this functionality ?

I don't understand the use-case for the MINP field, if we don't want the actual motors to stop when a soft-motor's hardware limit is activated.

tboegi commented:
The original motivation comes from a soft-slit database:
2 physical (hard) motors (the pos- and neg- blade)
controlled by 4 logical motors:
gap, center, pos, neg.
The MINP field would allow to 
a) forward limit switches: from pos+HLS -> gapHLS, centerHLS
  using a calc or transform record.
b) handle more bits (homed, power on) from hard- 
   to soft-motors.

The STOP function is  another nice feature.
I will look at it the next days.

mp49 commented:
If we add a STOP feature it should definitely be optional. Presumably the underlying motor controller or PLC will handle stopping on a limit switch. Additional software STOPs can cause issues with motion programs implemented under the feet of the motor record.

tboegi commented:
Yes, the more I think about it, the more I am tempted to say that a STOP function could and should be implemented in
external records, like calc or transform.
The MINP should just forward bits.
And those can come from other records, or directly from a hard-motor.

kmpeters commented:
I'll do some testing with an external stop calcout record.

kmpeters commented:
> I'll do some testing with an external stop calcout record.

This updated `SoftMSTA.db` works reasonably well with simulated motors:

```
record(motor, "$(P)$(SM)")
{
  field(MINP, "$(P)$(SM):msta NPP NMS")
}

record(mbboDirect, "$(P)$(SM):msta")
{
  field(DESC, "User-specified motor status")
  field(DTYP, "Soft Channel")
  field(OMSL, "supervisory")
  #field(OUT,  "TBD")
  field(SHFT, "0")
}

record(calcout, "$(P)$(SM):stopCalc")
{
  field(DESC, "HW limit - stop motor")
  field(INPA, "$(P)$(SM).MSTA CP NMS")
  field(CALC, "(A&(1<<2))||(A&(1<<13))")
  field(OOPT, "When Non-zero")
  field(DOPT, "Use OCAL")
  field(OCAL, "1")
  field(OUT,  "$(P)$(M).STOP PP NMS")
}
```
Every move that is made while the hardware limit of the soft motor is active is stopped, but the motor is able to move small amounts that would, in theory, allow moving off of the limit switch.

tboegi commented:
@kmpeters 
Do I understand that right:
The motorRecord.cc allow to move forward, when the HLS is active ?
That feels like a bug to me.

kmpeters commented:
Implementing hardware-limit enforcement in the soft-channel device support will likely require the direction bit to be set properly, which is also not implemented.

tboegi commented:
@kmpeters 
Kevin, we are slightly getting off track.
I created 
https://github.com/epics-modules/motor/issues/212

I will continue to work on both of these. Stay tuned.

tboegi committed:
Bugfix for MINP: it is a long

When implementing the MINP (MSTA input) it was overseen,
that MSTA is defined as 32 bit (where only 15 bits are used).
Correct this: use long instead of short.

tboegi committed:
motorRecord.html: Add the MINP field

Add the MINP field: It is "similar" to e.g. DINP

tboegi committed:
Merge 'motor/master' into torsten/231206-add-MINP-for-softmotor
